### ISSUE TITLE
telegram-stock-bot

### ISSUE THREAD

**Issue by Akmalzuhdi26** [Akmalzuhdi26/telegram-stock-bot#1]
import logging
import matplotlib.pyplot as plt
import io
import yfinance as yf
from datetime import datetime, timedelta, time as dtime
from telegram import Update, Bot, InputFile
from telegram.ext import ApplicationBuilder, CommandHandler, ContextTypes, JobQueue

# Konfigurasi logging
logging.basicConfig(format='%(asctime)s - %(name)s - %(levelname)s - %(message)s', level=logging.INFO)

# Token dari @BotFather
TOKEN = 'YOUR_TELEGRAM_BOT_TOKEN_HERE'
USER_ID = 6955102023  # ID Telegram @Denakmal26

# Daftar saham untuk dianalisis
saham_watchlist = ['DEWA', 'ANTM', 'BBRI', 'BRIS', 'MDKA']

# Fungsi membuat sinyal harian berdasarkan data yfinance
def generate_sinyal():
    hasil = {}
    for kode in saham_watchlist:
        ticker = yf.Ticker(kode + ".JK")
        data = ticker.history(period="15d")

        if data.empty:
            continue

        close = data['Close']
        volume = data['Volume']

        rsi = hitung_rsi(close)
        ma20 = close.rolling(window=20).mean()
        ma50 = close.rolling(window=50).mean()

        sinyal = 'WAIT'
        indikator = []

        if close.iloc[-1] > ma20.iloc[-1] > ma50.iloc[-1]:
            sinyal = 'BUY'
            indikator.append('MA20 > MA50')

        if rsi < 30:
            indikator.append('RSI Oversold')

        if rsi > 70:
            sinyal = 'SELL'
            indikator.append('RSI Overbought')

        hasil[kode] = {
            'sinyal': sinyal,
            'entry': f"{close.iloc[-1]:.2f}",
            'target': '-',
            'stop_loss': '-',
            'indikator': ', '.join(indikator) if indikator else 'Netral'
        }
    return hasil

# Fungsi hitung RSI sederhana

def hitung_rsi(series, periode=14):
    delta = series.diff()
    gain = delta.where(delta > 0, 0.0)
    loss = -delta.where(delta < 0, 0.0)
    avg_gain = gain.rolling(window=periode).mean()
    avg_loss = loss.rolling(window=periode).mean()
    rs = avg_gain / avg_loss
    rsi = 100 - (100 / (1 + rs))
    return rsi.iloc[-1] if not rsi.isna().iloc[-1] else 50

# Fungsi membuat pesan sinyal rapi
def buat_pesan_sinyal():
    sinyal_harian = generate_sinyal()
    pesan = "[SINYAL SAHAM HARI INI]\n"
    for kode, data in sinyal_harian.items():
        pesan += f"\n{kode}\nSinyal: {data['sinyal']}\nEntry: {data['entry']}\nTarget: {data['target']}\nStop Loss: {data['stop_loss']}\nIndikator: {data['indikator']}\n"
    return pesan

# Command handlers
async def start(update: Update, context: ContextTypes.DEFAULT_TYPE):
    await update.message.reply_text(
        "Selamat datang di Bot Sinyal Saham!\n"
        "Gunakan perintah berikut:\n"
        "/sinyal_hari_ini - Lihat sinyal saham terbaru\n"
        "/watchlist - Lihat daftar saham pantauan\n"
        "/chart KODESAHAM - Lihat grafik harga saham\n"
        "/harga KODESAHAM - Lihat harga terbaru"
    )

async def sinyal_hari_ini(update: Update, context: ContextTypes.DEFAULT_TYPE):
    pesan = buat_pesan_sinyal()
    await update.message.reply_text(pesan)

async def watchlist_hari_ini(update: Update, context: ContextTypes.DEFAULT_TYPE):
    pesan = "[DAFTAR WATCHLIST]\n" + '\n'.join([f"- {kode}" for kode in saham_watchlist])
    await update.message.reply_text(pesan)

async def harga(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if context.args:
        kode = context.args[0].upper()
        saham = yf.Ticker(kode + ".JK")
        info = saham.history(period="1d")
        if not info.empty:
            last_price = info['Close'].iloc[-1]
            await update.message.reply_text(f"Harga terakhir {kode}: Rp {last_price:,.2f}")
        else:
            await update.message.reply_text(f"Data untuk {kode} tidak ditemukan.")
    else:
        await update.message.reply_text("Gunakan format: /harga KODESAHAM")

async def chart(update: Update, context: ContextTypes.DEFAULT_TYPE):
    if context.args:
        kode = context.args[0].upper()
        saham = yf.Ticker(kode + ".JK")
        end_date = datetime.now()
        start_date = end_date - timedelta(days=30)
        data = saham.history(start=start_date, end=end_date)

        if data.empty:
            await update.message.reply_text("Data tidak ditemukan untuk kode saham tersebut.")
            return

        plt.figure(figsize=(10, 5))
        plt.plot(data.index, data['Close'], label='Harga Penutupan')
        plt.title(f"Grafik {kode} - 30 Hari Terakhir")
        plt.xlabel("Tanggal")
        plt.ylabel("Harga")
        plt.grid(True)
        plt.legend()
        buf = io.BytesIO()
        plt.savefig(buf, format='png')
        buf.seek(0)
        await update.message.reply_photo(photo=InputFile(buf, filename=f"chart_{kode}.png"))
    else:
        await update.message.reply_text("Gunakan format: /chart KODESAHAM")

# Scheduler harian
async def kirim_sinyal_otomatis(context: ContextTypes.DEFAULT_TYPE):
    pesan = buat_pesan_sinyal()
    await context.bot.send_message(chat_id=USER_ID, text=pesan)

# Main bot app
if __name__ == '__main__':
    app = ApplicationBuilder().token(TOKEN).build()

    app.add_handler(CommandHandler("start", start))
    app.add_handler(CommandHandler("sinyal_hari_ini", sinyal_hari_ini))
    app.add_handler(CommandHandler("watchlist", watchlist_hari_ini))
    app.add_handler(CommandHandler("harga", harga))
    app.add_handler(CommandHandler("chart", chart))

    # Jadwal otomatis jam 08:30 WIB (01:30 UTC)
    job_queue: JobQueue = app.job_queue
    job_queue.run_daily(kirim_sinyal_otomatis, time=dtime(hour=1, minute=30))

    print("Bot sedang berjalan...")
    app.run_polling()
